src/Microsoft.VisualStudio.Threading/NonConcurrentSynchronizationContext.cs

```diff
@@ -0,0 +1,178 @@
+
+            Requires.NotNull(d, nameof(d));
+
+            this.queue!.Enqueue((d, state));
+        }
+
+        /// <inheritdoc />
+        public override void Send(SendOrPostCallback d, object? state)
+        {
+            Requires.NotNull(d, nameof(d));
+
+            if (this.copyOf is object)
+            {
+                this.copyOf.Send(d, state);
+                return;
+            }
+
+            // Allow inlining if we're on the already-assigned thread.
+            if (this.activeManagedThreadId is int activeThread && Environment.CurrentManagedThreadId == activeThread)
+            {
+                using (this.sticky ? this.Apply(checkForChangesOnRevert: false) : default)
+                {
+                    d(state);
+                }
+
+                return;
+            }
+
+            var tcs = new TaskCompletionSource<object?>();
+            this.Post(
+                s2 =>
+                {
+                    var (cb, s, m) = (Tuple<SendOrPostCallback, object, TaskCompletionSource<object?>>)s2!;
+                    try
+                    {
+                        cb(s);
+                        m.SetResult(null);
+                    }
+                    catch (Exception ex)
+                    {
+                        m.SetException(ex);
+                    }
+                },
+                Tuple.Create(d, state, tcs));
+            tcs.Task.GetAwaiter().GetResult();
+        }
+
+        /// <inheritdoc />
+        public override SynchronizationContext CreateCopy() => new NonConcurrentSynchronizationContext(this);
+
+        /// <summary>
+        /// Executes queued work on the threadpool, one at a time.
+        /// </summary>
+        /// <returns>A task that always completes successfully.</returns>
+        [SuppressMessage("Microsoft.Globalization", "CA1303:Do not pass literals as localized parameters", MessageId = "System.Environment.FailFast(System.String,System.Exception)")]
+        private async Task ProcessQueueAsync()
+        {
+            try
+            {
+                while (!this.queue!.IsCompleted)
+                {
+                    var work = await this.queue.DequeueAsync().ConfigureAwait(false);
+                    this.activeManagedThreadId = Environment.CurrentManagedThreadId;
+                    try
+                    {
+                        using (this.sticky ? this.Apply(checkForChangesOnRevert: false) : default)
+                        {
+                            work.Item1(work.Item2);
+                        }
+                    }
+                    catch (Exception ex)
+                    {
+                        this.UnhandledException?.Invoke(this, ex);
+                    }
+                    finally
+                    {
+                        this.activeManagedThreadId = null;
+                    }
+                }
+            }
+            catch (OperationCanceledException) when (this.queue!.IsCompleted)
```

Comment:
Are we relying on gc to complete the queue? I noticed some sync contexts like WIndowsForms one is Disposable, would it make sense to make this disposable too?

Reply:
In my opinion SC should not be IDisposable, it is not part of the SC contract, people interacting with the base class will not intentionally Dispose but may accidently do so.

SC's are supposed to be lightweight, you need to be able to spawn cheap copies which refer to a SC from a separate thread (some part of the framework uses reference identity to determine whether it is possible to avoid posting to the SC and inline a delegate call, I happened to write about that before on this repo so it should be possible to find in a search).

When I write new SC implementations I create a controlling object implementing IDisposable and allow creating SC instances which know how to post callbacks to their controlling object. You shut down the controlling object and not the SC instances (what would calling Dispose on a copy of a SC even mean?)

Reply:
> would it make sense to make this disposable too?

The internal class that this one is based on _was_ actually disposable. I removed it from this class because it's hard to reason about it -- who owns the SynchronizationContext? And what happens when `CreateCopy` is called and the copy is disposed of? We weren't even ever disposing the disposable one this was based on, so I decided to remove that.
I guess I can remove the completion testing from this code though, since it is never completed.

> (some part of the framework uses reference identity to determine whether it is possible to avoid posting to the SC and inline a delegate call, I happened to write about that before on this repo so it should be possible to find in a search).

Yup! That's in our `SingleThreadedSynchronizationContext` class. I reviewed that this morning before writing this code to make sure I didn't re-make any mistakes and I saw that comment and couldn't remember how we came across that knowledge. Thanks for sharing, @weltkante!

Regarding GC: yes any of these are subject to garbage collection once all references go away. Even an async state machine will evaporate if no one could ever cause it to run again. So it works out.